odoo/models.py

```diff
@@ -3141,103 +3171,116 @@ def _fetch_field(self, field):
                 fnames.append(field.name)
         else:
             fnames = [field.name]
-        self._read(fnames)
-
-    def _read(self, field_names):
-        """ Read the given fields of the records in ``self`` from the database,
-            and store them in cache. Skip fields that are not stored.
+        self.fetch(fnames)
 
-            :param field_names: list of field names to read
+    def fetch(self, field_names):
+        """ Make sure the given fields are in memory for the records in ``self``,
+        by fetching what is necessary from the database.  Non-stored fields are
+        mostly ignored, except for their stored dependencies.
         """
-        if not self:
+        if not self or not field_names:
             return
-        self.check_access_rights('read')
 
-        # determine columns fields and those with their own read() method
-        column_fields = []
-        other_fields = []
-        translated_field_names = []
-        for name in field_names:
-            if name == 'id':
-                continue
-            field = self._fields.get(name)
+        # determine fields to fetch
+        fields_to_fetch = OrderedSet()
+        cache = self.env.cache
+        field_names = self.check_field_access_rights('read', field_names)
+        for field_name in field_names:
+            field = self._fields.get(field_name)
             if not field:
-                _logger.warning("%s._read() with unknown field %r", self._name, name)
+                raise ValueError("Invalid field %r on model %r" % (field_name, self._name))
+            if not any(cache.get_missing_ids(self, field)):
                 continue
-            if field.base_field.store and field.base_field.column_type:
-                column_fields.append(field)
-            elif field.store and not field.column_type:
-                # non-column fields: for the sake of simplicity, we ignore inherited fields
-                other_fields.append(field)
-            if field.store and field.translate:
-                translated_field_names.append(field.name)
+            if field.store:
+                fields_to_fetch.add(field)
+            elif field.compute:
+                # optimization: fetch direct field dependencies
+                for dotname in self.pool.field_depends[field]:
+                    dep = self._fields[dotname.split('.', 1)[0]]
+                    if dep.prefetch is True and (not dep.groups or self.user_has_groups(dep.groups)):
+                        fields_to_fetch.add(dep)
+
+        if not fields_to_fetch:
+            # there is nothing to fetch, but we expect an error anyway in case
+            # self is not accessible
+            self.check_access_rights('read')
+            try:
+                self.check_access_rule('read')
+            except MissingError:
+                self.exists().check_access_rule('read')
```

Comment:
I think we should leave that to another task, which would revamp the API and implementation of access rights...  I will make both copies of that code identical, and as they are only a few lines apart, it will be easy to refactor.